core/modules/filter/filter.pages.inc

```diff
@@ -491,28 +491,169 @@ function filter_format_editor_dialog_save($form, &$form_state) {
 
-  // If Link module is available, automatically clean up URL, automatically
-  // adding http:// or a prefixing slash if needed.
-  if (module_exists('link') && $type = link_validate_url($value)) {
-    $value = link_cleanup_url($value);
-    $form_state['values']['type'] = $type;
-    if ($type === LINK_INTERNAL && strpos($value, '/') !== 0) {
-      $value = '/' . $value;
-    }
+  // Allow empty values, since this will not result in a link tag anyway.
+  if ($value === '') {
+    form_set_value($element, $value, $form_state);
+    return;
   }
 
+  // Cleanup the URL so it's more likely to pass validation.
+  $value = _filter_cleanup_url($value);
+
   form_set_value($element, $value, $form_state);
 
+  // If external URL ensure it doesn't have dangerous protocols.
+  $dangerous_protocols = _filter_check_dangerous_protocols($value);
+
   // Unlike #type = 'url' validation, we allow both relative and absolute paths.
-  if ($value !== '' && !valid_url($value, TRUE) && !valid_url($value, FALSE)) {
+  if ($dangerous_protocols || (!valid_url($value, TRUE) && !valid_url($value, FALSE))) {
     form_error($element, t('The URL %url is not valid.', array('%url' => $value)));
   }
 }
 
+/**
+ * Clean up URL for validation.
+ *
+ * Attempts to create a URL that will pass validation: adds a protocol if it's
+ * missing; urlencodes an anchor link; attempts to find the proper multilingual
+ * URL; and urlencodes any internal Backdrop path.
+ *
+ * @param string $url
+ *
+ * @return string
+ *   A cleaned-up URL that is either a properly formatted external URL or an
+ *   internal path.
+ *
+ * @see _filter_format_editor_link_url_validate()
+ */
+function _filter_cleanup_url($url) {
+  // Will add a default protocol if missing on an external url.
+  $url = _filter_create_valid_external_url($url);
+
+  if (_filter_url_is_external($url)) {
+    return $url;
+  }
+
+  // Urlencode an anchor as it's more likely to be typed in with unicode.
+  if (substr($url, 0, 1) == '#' && preg_match('/^(#)([\S]+)$/', $url, $matches)) {
+    return $matches['1'] . backdrop_encode_path($matches['2']);
+  }
+
+  // Break up the components so we can build up an aliased internal path.
+  $url_parts = backdrop_parse_url($url);
+
+  // Lookup valid internal paths if multilingual
+  if (module_exists('locale') && language_multilingual()) {
+    if ($language = _filter_lookup_locale_path($url)) {
+      $url_parts['language'] = $language;
+    }
+  }
+
+  return url($url_parts['path'], $url_parts);
+}
+
+/**
+ * Forms a valid external URL if possible.
+ *
+ * Trims whitespace and automatically adds an http:// to addresses that
+ * appear to be external URLs but have no specified protocol.
+ *
+ * @param string $url
+ *   The URL entered by the user.
+ *
+ * @return string
+ *   The URL with the protocol prepended if needed.
+ *
+ * @see link_cleanup_url().
+ */
+function _filter_create_valid_external_url($url) {
+  $url = trim($url);
+
+  // Check if there is no protocol specified.
```

Comment:
This definitely breaks protocol-relative urls (which would be valid).

Reply:
I borrowed that regex from the link module. I can't recall which function offhand. I'm going to take another look at this regex.

Reply:
I got it from `link_cleanup_url()` (I remembered I put it right in the docblock).

Do you have an example of a protocol relative URL? Do you mean one like `//archive.org`? With that example it isn't matched in the first regex and it also isn't matched in the second. That's why it doesn't get an http added. Even though it doesn't think it has a protocol, it also doesn't think it starts with a domain.